<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<title></title>
</head>
<body>
	<h2>aCGH tools / Call copy number aberrations</h2>
	
	<h3>Description</h3>
	<p>Call copy number aberrations from aCGH log ratios.</p>
	
	<h3>Parameters</h3>
	<ul>
		<li>Normalization method (median, mode, none) [none]</li>
		<li>Number of chromosomes (...) [23]</li>
		<li>Number of copy number states (3, 4) [3]</li>
		<li>Minimum number of probes per segment (2, 3, 4, 5) [2]</li>
		<li>Minimum number of SDs between segments [0.0]</li>
		<li>Organism</li>
		<li>Human genome build</li>
	</ul>
	
	<h3>Details</h3>
	<p>Log ratios are first segmented using DNAcopy. Segments are consecutive areas that most likely share the same copy number and are separated by breakpoints. Discrete copy number levels are then called with CGHcall. Calling of three copy number states means separating losses, normals and gains. Amplifications can also be detected separately as the fourth copy number state. For each probe, the frequencies of arrays exhibiting a loss/gain/amplification are also included.</p>
	<p>The input data must contain these columns: chromosome, start, end.</p>
	
	<h3>Output</h3>
	<p>The inputted log ratios are accompanied with segmented log ratios, hard copy number calls, and probabilities assigned to each call.</p>
	<p>A plot is also returned that shows the frequencies of gains (blue) and losses (red).</p>
	
	<h3>References</h3>
	<p>DNAcopy:<br />
		Venkatraman et al. (2007) A faster circular binary segmentation algorithm for the analysis of array CGH data. Bioinformatics 23: 657-663</p>
	<p>CGHcall:<br />
		Wiel et al. (2007) CGHcall: calling aberrations for array CGH tumor profiles. Bioinformatics 23: 892-894</p>
</body>
</html>
