<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Matching genomic regions / Overlap BEDPE with BED</h2>

<h3>Description</h3>

<p>
Report overlaps between a BEDPE file and a BED/GFF/VCF file.
</p>

<h3>Parameters</h3>

<ul>
<li>File A is BAM format (yes, no) [no]</li>
<li>Write uncompressed BAM output (yes, no) [no]</li>
<li>When using BAM input, write output as BEDPE (yes, no) [no]</li>
<li>Use BAM total edit distance for BEDPE score (yes, no) [no]</li>
<li>Minimum overlap (0.0-1.0) [0.000000001]</li> 
<li>Force strandedness (yes, no) [no]</li>
<li>Approach to reporting overlaps (either, neither, both, xor, notboth, ispan, ospan, notispan, notospan) [either]</li>
</ul>

<h3>Details</h3>

<p>
pairToBed compares each end of a BEDPE feature or a paired-end BAM alignment to a feature file in
search of overlaps.
</p>
<p>
NOTE: pairToBed requires that the BAM file is sorted/grouped by the read name. This
allows pairToBed to extract correct alignment coordinates for each end based on their
respective CIGAR strings. It also assumes that the alignments for a given pair come in
groups of twos. There is not yet a standard method for reporting multiple alignments
using BAM. pairToBed will fail if an aligner does not report alignments in pairs. 
</p>
<p>
Selecting "File A is BAM format [yes]" is analogous to using "-abam" option instead of "-a" in the command line version.
</p>

<h3>Output</h3>

<p>
The output is either a BEDPE file or a compressed or uncompressed BAM file depending on input file and options selected.
</p>

<h3>References</h3>

<p>This tool uses the BEDTools package. Please cite the article:</p>

<p>Quinlan AR and Hall IM, 2010. <a href="http://bioinformatics.oxfordjournals.org/content/26/6/841.abstract?sid=62330781-9bde-4e76-8f13-b5e323cf8034">BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841–842.</a> </p>

<p>Please see the <a href = "http://code.google.com/p/bedtools/">BEDTools homepage</A> for more details.</p>
</body>

