<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Matching genomic regions / Genome coverage BED</h2>

<h3>Description</h3>

<p>
Compute the coverage of a feature file among a genome.
</p>

<h3>Parameters</h3>

<ul>
<li>Input file is BAM format (yes, no) [no]</li>
<li>Output type (histogram, depth, BedGraph, BedGraph-all) [histogram]</li>
<li>Treat split BAM or BED12 entries as distinct BED intervals (yes, no) [no]</li>
<li>Calculate coverage from specific strand (both, +, -) [both]</li> 
<li>Combine positions by depth (yes, no) [no]</li>
<li>Max depth (Integer) [1]</li>
</ul>

<h3>Details</h3>

<p>
genomeCoverageBed computes a histogram of feature coverage (e.g., aligned sequences) for a given
genome. Optionally, by using the "Output type [depth]" option, it will report the depth of coverage at each base on each
chromosome in the genome file.
</p>
<p>
Selecting "File A is BAM format [yes]" is analogous to using "-abam" option instead of "-a" in the command line version.
</p>

<h3>Output</h3>

<p>
The output is a tab-delimited text file.
</p>

<h3>References</h3>

<p>This tool uses the BEDTools package. Please cite the article:</p>

<p>Quinlan AR and Hall IM, 2010. <a href="http://bioinformatics.oxfordjournals.org/content/26/6/841.abstract?sid=62330781-9bde-4e76-8f13-b5e323cf8034">BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841–842.</a> </p>

<p>Please see the <a href = "http://code.google.com/p/bedtools/">BEDTools homepage</A> for more details.</p>
</body>

