<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Matching genomic regions / Compare two BEDPE files</h2>

<h3>Description</h3>

<p>
Report overlaps between two paired-end BED files (BEDPE).
</p>

<h3>Parameters</h3>

<ul>
<li>Minimum overlap (0.0-1.0) [0.000000001]</li> 
<li>Approach to reporting overlaps (either, neither, both) [both]</li> 
<li>Ignore strands when searching for overlaps (yes, no) [no]</li>
<li>Require different names (yes, no) [no]</li>
<li>Add slop (yes, no) [no]</li>
<li>The amount of slop in bp to be added to each footprint (Integer) [0]</li> 
<li>Add slop based to each BEDPE footprint based on strand (yes, no) [no]</li>
</ul>

<h3>Details</h3>

<p>
pairToPair compares two BEDPE files in search of overlaps where each end of a BEDPE feature in A
overlaps with the ends of a feature in B. For example, using pairToPair, one could screen for the exact
same discordant paired-end alignment in two files. This could suggest (among other things) that the
discordant pair suggests the same structural variation in each file/sample.
</p>

<h3>Output</h3>

<p>
The output is a BEDPE file.
</p>

<h3>References</h3>

<p>This tool uses the BEDTools package. Please cite the article:</p>

<p>Quinlan AR and Hall IM, 2010. <a href="http://bioinformatics.oxfordjournals.org/content/26/6/841.abstract?sid=62330781-9bde-4e76-8f13-b5e323cf8034">BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841–842.</a> </p>

<p>Please see the <a href = "http://code.google.com/p/bedtools/">BEDTools homepage</A> for more details.</p>
</body>

