<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Utilities / Convert BAM to BED</h2>

<h3>Description</h3>

<p>
Converts BAM alignments to BED6, BED12 or BEDPE format.
</p>

<h3>Parameters</h3>

<ul>
<li>Output type (BED6, BED12, BEDPE) [BED6]</li>
<li>Report split BAM alignments as separate BED entries (yes, no) [no]</li>
<li>Use BAM edit distance for BED score (yes, no) [no]</li>
<li>Color string (0-255,0-255,0-255) [255,0,0]</li>
<li>Add the CIGAR string (yes,no) [no]</li>
</ul>

<h3>Details</h3>

<p>
bamToBed is a general purpose tool that will convert sequence alignments in BAM format to either
BED6, BED12 or BEDPE format. This enables one to convert BAM files for use with all of the other
BEDTools. The CIGAR string is used to compute the alignment end coordinate in an “ungapped”
fashion. That is, match (“M”), deletion (“D”), and splice (“N”) operations are observed when computing
alignment ends.
</p>
<p>
Note that when selecting BEDPE format, it is required that the BAM file is sorted/grouped by
the read name. This allows bamToBed to extract correct alignment coordinates for each
end based on their respective CIGAR strings. It also assumes that the alignments for a
given pair come in groups of twos.
</p>

<h3>Output</h3>

<p>
The output is a BED6, BED12 or BEDPE format depending on the selected parameter.
</p>

<h3>References</h3>

<p>This tool uses the BEDTools package. Please cite the article:</p>

<p>Quinlan AR and Hall IM, 2010. <a href="http://bioinformatics.oxfordjournals.org/content/26/6/841.abstract?sid=62330781-9bde-4e76-8f13-b5e323cf8034">BEDTools: a flexible suite of utilities for comparing genomic features. Bioinformatics. 26, 6, pp. 841–842.</a> </p>

<p>Please see the <a href = "http://code.google.com/p/bedtools/">BEDTools homepage</A> for more details.</p>
</body>

