<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Clustering / Hierarchical</h2>

<h3>Description</h3>

<p>Hierarchical clustering creates a dendrogram describing the relationships between genes or chips in a selected
genelist. Clustering consists of two separate steps. First, all pairwise distances between objects (genes or chips)
are calculated. The dendrogram is then drawn using these distances by a selected method. The maximum number of genes/samples to be clustered is 
20 000.
</p>

<h3>Parameters</h3>

<ul>
<li>What to cluster (genes, chips) [genes]</li>
<li>Distance method (euclidian, manhattan, pearson, spearman) [euclidian]</li>
<li>Tree method (single, average, complete, ward) [average]</li>
<li>Resampling method (none, permutation-topodist, bootstrap) [none]</li>
<li>Number of replications (1-10000) [1000]</li>
<li>Image width (200-3200) [600]</li>
<li>Image height (200-3200) [600]</li>
</ul>

<h3>Details</h3>


User can select from four different distance methods:
<ul>
<li>Euclidean distance</li>
  <dl><dd>This is an absolute distance between two gene expression profiles or chips. Euclidean distances between
      objects, such as dots on a paper could be measured with a ruler.</dd></dl>
<li>Manhattan distance</li>
  <dl><dd>A distance between gene expression profiles or chips measured at right angles.</dl></dd>
<li>Pearson correlation</li>
  <dl><dd>Distance using Pearson correlation is calculated as 1-correlation coefficient.</dl></dd>
<li>Spearman correlation</li>
  <dl><dd>Distance using Spearman correlation is calculated as 1-correlation coefficient.</dl></dd>
</ul>

It is also possible to modify the dendrogram drawing method. Four possible options are:
<ul>
<li>Single linkage</li>
  <dl><dd>A distance between clusters in the tree is calculated using the shortest distance between them.</dl></dd>
<li>Average linkage (UPGMA)</li>
  <dl><dd>A distance between clusters in the tree is calculated using average distance between them.</dl></dd>
<li>Complete linkage</li>
  <dl><dd>A distance between clusters in the tree is calculated using the longest distance between them.</dl></dd>
<li>Ward</li>
  <dl><dd>At every step of clustering two clusters that result into a minimal loss of information are combined.
          Information loss is measured using error sum-of-squares criterion.</dl></dd>
</ul>


<p>The results of the hierarchical clustering can be checked using bootstrapping testing. Bootstrapping
creates a user specified number of pseudodatasets from the original one. In the pseudodatasets, each row or column (depending
on whether genes or chips were clustered), can be present zero, one or several times. Every bootstrapped dataset is then
converted into a dendrogram. Say, if 100 bootstrap samples were used, 100 trees are produced. A majority rule consensus
is then created from these trees, and the results are displayed for the user. In the majority rule consensus tree every
node in the tree is labeled with a number. This number represents the number of trees where that node was present. The higher
the number the better. Note that bootstrapping can only be done using Euclidian distance or Pearson correlation.
Bootstrap resampling on datasets larger than 1000 genes is not possible due to computing time limitations. 
Please note that you can run hierarchical clustering on datasets including up to 20000 genes, provided the resampling option is turned off.
</p>

<h3>Output</h3>

<p>A file with information on how to draw the tree. This file can be visualised using the interactive "Hierarchical clustering" visualisation. 
Please note that you can select genes in this visualisation by drawing a box in the heatmap area. 
Clicking on the tab "Selected" allows you to create a new data set based on your selection.</p>
 
<h3>References</h3>

<p>This tools uses R packages ape and amap. The citation for ape is: </p>

<p>Paradis E., Claude J. & Strimmer K. 2004. APE: analyses of phylogenetics and evolution in R language. Bioinformatics 20: 289–290. PDF [37 KB].</p>

</body>

