<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
	<title></title>
</head>
<body>
	<h2>CNA-seq / Define CNA-seq experiment</h2>
	
	<h3>Description</h3>
	<p>This tool takes as input a number of files with binned read counts, combines them into a single table and generates the phenodata table required for further analyses.</p>
	
	<h3>Parameters</h3>
	<ul>
		<li>counts (original raw counts, GC corrected counts) [original raw counts]</li>
		<li>log2 transform counts (no, yes) [no]</li>
		<li>minimum mappability (0...1) [0]</li>
		<li>normalization (none, median) [none]</li>
		<li>organism (human) [human]</li>
		<li>human genome build (GRCh37/hg19) [GRCh37/hg19]</li>
	</ul>
	
	<h3>Details</h3>
	<p>The parameter values should be chosen based on which one of the following options is chosen for subsequent segmentation and calling of copy number aberrations. If the tool CNA-seq / Segment and call copy number aberrations is used, it contains its own GC correction and mappability filtering, so these should not be used and the data should not be normalized here. (counts: original raw counts, log2 transform counts: no, minimum mappability: 0, normalization: none)</p>
	<p>If segmentation and calling are performed with aCGH methods from the Microarrays tab, GC correction, log2 transformation, mappability filtering, and normalization should all be performed. (counts: GC corrected counts, log2 transform counts: yes, minimum mappability: e.g. 0.85, normalization: median)</p>
	
	<h3>Output</h3>
	<p>Data table with read counts per sample and accompanying phenodata table.</p>
	
	<h3>References</h3>
	<p>GC correction: Miller et al. (2011) ReadDepth: a parallel R package for detecting copy number alterations from short sequencing reads. PLoS One 6: e16327</p>
	<p>Mappability: Koehler et al. (2011) The uniqueome: a mappability resource for short-tag sequencing. Bioinformatics 27: 272-274</p>
</body>
</html>
