<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>RNA-seq /  Differential expression analysis using DESeq</h2>

<h3>Description</h3>

This tool will perform an analysis for differentially expressed sequences using the R implementation of the DESeq algorithm.

<br>
<br>
<h3>Parameters</h3>

<ul>
<li> Column describing groups [group]</li>
<li> Apply normalization (yes, no) [yes]</li>
<li> Disregard replicates (yes, no) [no]</li>
<li> Dispersion method (fit all, fit low) [fit all]</li>
<li> Dispersion estimate (parametric, local) [local] </li>
<li> Multiple testing correction (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
<li> P-value cutoff (0-1) [0.05]</li>
<li> Plot width (200-3200 [600]</li>
<li> Plot height (200-3200) [600]</li>
</ul>

<h3><br>Details</h3>

<br>
Given an input table of counts data for at least two samples, the DESeq package performs normalization, dispersion model fitting
and statistical analysis to identify differentially expressed genomic features between two experimental conditions.
Even though it is possible to run the analysis without independent biological replicates, by either using replicates
for only one experiment condition or by estimating variability between samples of different experiment conditions,
it is highly recommended to always include at least one additional biological replicate for each experiment condition.
<br><br>
Notice that in it's current implementation, the tool only supports single-factor experiment designs. The experiment
conditions to be compared should be defined in the phenodata.tsv file and the appropriate column be selected using
the 'Column describing group' parameter.
<br><br>
If normalization is enabled, scaling of the data to account for variations in library size, are accomplished either by dividing with the average of the total counts for each sample or,
if the user have filled in the 'library_size' column of the phenodata.tsv file, by dividing with the average of those values.
<br><br>
Dispersion values for each genomic feature is estimated through a model fit procedure, which can be performed in a "local" or "parametric" mode.
The former is selected by default, since it is more robust, but users are encouraged to experiment with the setting to optimize results.
The data is fitted in either of two ways, where the "fit all" option is the default choice and more conservative, whereas the "fit low" setting
may yield increased sensitivity but potentially at the cost of increased false positives. 
<br><br>
Statistical testing is performed using a generalized linear model and the p-values are adjusted for multiple testing using the classical approaches.
<br><br>

<h3>Output</h3>

<p>The analysis output consists of the following:</p>

<ul>
<li>de-list-deseq.tsv:	Table containing the results of the statistical testing, including fold change estimates and p-values.<br></li>
<li>de-list-deseq.bed:  The BED version of the results table contains genomic coordinates and p-value estimates and is ideal for quick navigation in the Genome Browser.<br></li>
<li>ma-plot-significant-deseq.pdf: A scatter plot where the significantly differentially expressed features are highlighted. <br></li>
<li>dispersion-plot.pdf: A plot that displays the dispersion estimates as a function of the counts values, with the fitted model overlaid. <br></li>
<li>p-value-plot-edger.pdf: Plot of the raw and adjusted p-value distributions of the statistical test. <br></li>
</ul>

<h3><br>References</h3>

<p>This tool uses the DESeq package for statistical analysis. Please read the following article for more detailed information:</p>

<p> S Anders and W H. <a href="http://genomebiology.com/2010/11/10/R106">Differential expression analysis for sequence count data. Genome Biology 2010, 11:R106.</a></p> 


<p> </a></p> 

</body>
