<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>RNA-seq /  Differential expression analysis using DESeq</h2>

<h3>Description</h3>

This tool performs differential expression analysis using the DESeq Bioconductor package.

<br>
<br>
<h3>Parameters</h3>

<ul>
<li> Column describing groups [group]</li>
<li> Apply normalization (yes, no) [yes]</li>
<li> Disregard replicates (yes, no) [no]</li>
<li> Use fitted dispersion values (when higher than original values, always) [when higher than original values]</li>
<li> Dispersion estimate (parametric, local) [local] </li>
<li> Multiple testing correction (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
<li> P-value cutoff (0-1) [0.05]</li>
<li> Plot width (200-3200 [600]</li>
<li> Plot height (200-3200) [600]</li>
</ul>

<h3><br>Details</h3>

<br>
Given an input table of raw counts, the DESeq package performs statistical analysis to identify differentially expressed genes or other genomic features between two experimental conditions.
Note that in its current implementation, the tool only supports single-factor experiment designs. The experiment
conditions to be compared should be defined in the phenodata.tsv file and the appropriate column selected using
the 'Column describing group' parameter.
<br><br>
When normalization is enabled, size factors are calculated by summing the counts for each sample, or 
using the library size given by the user in the phenodata.tsv. The former allows to correct for RNA composition bias
 (which can arise for example when only a small number of genes are very highly expressed in one experiment condition but not in the other). 
<br><br>
A dispersion value is estimated for each gene through a model fit procedure, which can be performed in a "local" or "parametric" mode.
The former is more robust, but users are encouraged to experiment with the setting to optimize results.
Users can select to replace the original dispersion values by the fitted ones always, or only when the fitted value is higher than the original one 
(more conservative option).
<br><br>
It is highly recommended to always have at least two biological replicates for each condition.
If this is not possible, you can run the analysis using replicates
for only one condition, or by estimating variability using samples of the two different conditions.
<br><br>
Statistical testing is performed using a negative binomial test.

<h3>Output</h3>

<p>The analysis output consists of the following files:</p>

<ul>
<li>de-list-deseq.tsv: Table containing the results of the statistical testing, including fold change estimates and p-values.<br></li>
<li>de-list-deseq.bed: The BED version of the results table contains genomic coordinates and log2 fold change values.<br></li>
<li>ma-plot-significant-deseq.pdf: A scatter plot where the significantly differentially expressed features are highlighted. <br></li>
<li>dispersion-plot.pdf: A plot that displays the dispersion estimates as a function of the counts values, with the fitted model overlaid. <br></li>
<li>p-value-plot-edger.pdf: Plot of the raw and adjusted p-value distributions of the statistical test. <br></li>
</ul>

<h3><br>References</h3>

<p>This tool uses the DESeq package for statistical analysis. Please read the following article for more detailed information:</p>

<p> S Anders and W H. <a href="http://genomebiology.com/2010/11/10/R106">Differential expression analysis for sequence count data. Genome Biology 2010, 11:R106.</a></p> 

</body>
