
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en"><head><title>Embster manual</title></head><body>

<h2>23.2.2011 What's new?</h2>
	
<ol>
<li><b>New analysis tools</b>
	<ul>
	  <li> <a href="http://www.csc.fi/english/research/sciences/bioscience/programs/hmmer/index_html">HMMER3</a> package contains tools for 
	  creating and using Hidden Markov Models (HMM). HMMs can be used to describe and analyze related or similar sequence areas. 
	  They are normally derived from multiple sequence alignments and contain position specific information about 
	  the probabilities of having certain nucleotides or amino acids in each position of an alignment.
	  The following tools are now available in Embster's HMM category:
	  <ul>
	  	<li>hmmalign: Align sequences to a profile HMM</li>
	  	<li>hmmbuild: Profile HMM construction from multiple sequence alignment</li>
	  	<li>hmmemit: Sample sequences from a profile HMM</li>
	  	<li>hmmfetch: Retrieve profile HMM models from Pfam databases based on their Pfam accession numbers</li>
	  	<li>hmmscan: Search sequences against a Pfam-database</li>
	  	<li>hmmsearch: Search one or more profiles against a sequence database</li>
	  <	li>hmmstat: Display summary statistics for a profile file</li>
	 </ul>
	<li><u>Retrieve data/ Retrieve sequences from database</u>: Given a list of sequence names or accession numbers, this tool fetches the sequences from 
	a database (EMBL, UniProt, PDB, Ensembl) and puts them together in a new fasta file (if you would like each sequence to be put into a separate file,
	please use the menu "File / Import from / Database" instead).</li>
   <li><u>Retrieve data/ Retrieve sequences from dataset</u>: Given a list of sequence names or accession numbers, this tool retrieves the sequences from 
	a user-specified sequence file and puts them together in a new fasta file.</li>
	</ul>

<li><b>Changes to general functionality</b>
	<ul>
		  	<li>.seq files can be viewed as text</li>
</ul>

</ol>
<hr>


</body></html>