<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>miRNA-seq / Correlate miRNA with target expression</h2>

<h3>Description</h3>

<p>Given a list of miRNA:s and a matching data set with gene expression data from a multi-condition or time-series experiment,
 this tool identifies which of the target genes that correlate positively or negatively with the miRNA expression.
 Note that the miRNA and gene expression data sets must have the same number of conditions or time-points.</p>

<h3>Parameters</h3>

<ul>
<li>Order column miRNA [EMPTY] </li>
<li>Order column gene [EMPTY</li>
<li>ID type (probe_id, entrez_id) [probe_id]</li>
<li>Correlation method (pearson, spearman, kendall) [pearson]</li>
<li>P-value threshold (0..1) [0.05]</li>
<li>P-value adjustment method (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
</ul>

<h3>Details</h3>

<p>This tool integrates expression data from matching multi-condition or time-course data sets with miRNA and gene
expression, and performs a correlation analysis to identify which of the gene targets exhibit significant positive
or negative correlation with the miRNA expression. In order to be able to match up the different conditions, or time-points, in the two data sets
a column that specifies the sample order has to be added to the phenodata file of each data set. It is advisable not to use the "group" column for this purpose,
but rather make an additional column titled "order". NOTE: you need to select both the file with the miRNA expression and the file with the matching gene expression,
which is done by holding down the Control key while mousing over the files in the "Datasets" or "Workflow" windows.<br><br>The analysis consists of two parts, that perform the following:<br><br>Firstly, predicted target genes for the list of miRBA:s are fetched from the targetScan and PicTar databases, after which an intersection
is made so as to include only those genes that are predicted in both databases. Then, another intersection is made with the list of genes
present in the matching gene expression data set.<br><br>Secondly, the correlation is calculated between the expression of a miRNA and each of its' target genes, using Pearson, Spearman or Kendall's measure of correlation.
Pearson correlation is parametric, so can be sensitive to outliers, while the rank-based Spearman method is not and therefore more robust. In cases when the direction
of the change between two-conditions, or time-points, is more important than the absolute values of the expression profile, Kendall's correlation method is recommended.

The statistically significant results are reported in separate tables for the positively and negatively correlating genes, respectively.  
  
</p>

<h3>Output</h3>

<p>Output consists of two text files, one listing the significantly positevly correlating target genes and one listing the significantly negatively correlating ones, .</p>

</p>

<h3>References</h3>
<p>This tool leverages some of the functionality from the RmiR Bioconductor package. More info on the databases of predicted miRNA targets can be found at:<br/>
<br/>targetScan<br/>
http://www.targetscan.org/
<br/><br> PicTar<br/>http://pictar.mdc-berlin.de/</p>



</body>

