<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Pathways / SAFE</h2>

<h3>Description</h3>

<p>Searches for over- or under-represented genes for KEGG categories.</p>

<h3>Parameters</h3>

<ul>
<li>Phenodata column to test (...) </li>
<!--<li>Which ontology to test (KEGG) [KEGG]</li>-->
<li>Minimum category size (1...100) [10]</li>
<li>P value threshold (0-1) [0.05] </li>
<li>Image width (200...3200) [600]</li>
<li>Image height (200...3200) [600]</li>
</ul>

<h3>Details</h3>

<p>This tool takes a filtered gene list, and uses information about the genes, their expression and
grouping (from the phenodata group-column) to estimate the over-representation of the gene in the
categories. As such, this method is very similar to GSEA analysis from the Broad Institute.</p>

<!--<p>Note that calculations with GO categories can be very slow.</p>-->

<h3>Output</h3>

<p>Output is an text file with a list of significant categories, and a safe plot for the most significant
category.</p>

<h3>References</h3>

<p>If you use this tool, please cite:</p>

<p>Barry,W.T., Nobel, A.B. and Wright, F.A. (2005): <a href="http://bioinformatics.oxfordjournals.org/cgi/content/full/21/9/1943">Significance Analysis of
functional categories in gene expression studies: a structured permutation
approach</a>, Bioinformatics, 21(9):1943-1949.
</p>

</body>

