<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Cosmo</h2>

<h3>Description</h3>

<p>This tools retrieves promoter sequences for a set of genes and finds shared sequence motifs in them. Currently the tool works only for human, mouse, rat, drosophila and yeast genes.</p>

<h3>Parameters</h3>

<ul>
<li>Species (human, mouse, rat, drosophila, yeast) [human]</li>
<li>Promoter size (small, medium, large) [small]</li>
<li>Analyze strands (single, both) [single] </li>
<li>Motif appears more than once per sequence (yes, no) [no]</li>
<li>Percentage of sequences the motif should appear in (1-100) [50]</li>
<li>Transcription factor binding site size (small, medium) [small]</li>
</ul>

<h3>Details</h3>


<p>This tool retrieves upstream sequences for the specified genes and submits them to the Cosmo program. It needs to access the
chip-specific annotations, so if you have not specified the chiptype
during normalization of, e.g., Illumina data, it will not work. RefSeq IDs are used for retrieving 
promoter sequences constructed and annotated by UCSC genome browser staff. The same promoter sequences can be downloaded 
as a single FastA-formatted file from <a href="http://hgdownload.cse.ucsc.edu/downloads.html">UCSC Golden Path folder</a>. 
User can define how long promoter sequences are used for the analysis:</p>

<pre>
		Human		Mouse		Rat		Drosophila	Yeast
Small		1000 bp		1000 bp		1000 bp		1000 bp 	500 bp
Medium		2000 bp		2000 bp		2000 bp		2000 bp		1000 bp
Large		5000 bp		5000 bp		5000 bp		5000 bp		2500 bp
</pre>

<p>After retrieving the promoter sequences, they are submitted to the Cosmo program that finds common motifs
(putative transcription factor binding sites). Cosmo can find motifs of two different sizes: small (6-8 nucleotides) and large (10 nucleotides).
In addition, the user can define whether both strands of DNA should be searched for the common motifs, whether the motif should appear more than once in a sequence, and what proportion of the sequences should contain the motif.</p>

<h3>Output</h3>

<p>Cosmo returns common motifs written to a text file. These can then be further analyzed to 
verify whether they are known transcription factor binding sites. Cosmo also returns an image file containing a sequence logo representation of the motif, and another image showing the location of the motifs. </p>

<h3>References</h3>

<p>This tool uses the <a href="http://www.bioconductor.org/packages/bioc/vignettes/cosmo/inst/doc/cosmo.pdf">Bioconductor package Cosmo</a>. Please cite the article:</p>

<p>O. Bembom, S. Kele¸s, and M.J. van der Laan. Supervised detection of conserved motifs in
DNA sequences with cosmo. Statistical Applications in Genetics and Molecular Biology:
Vol. 6 : Iss. 1, Article 8. URL http://www.bepress.com/sagmb/vol6/iss1/art8.</p>

<h3>Notes</h3>

<p>The use of this tool is allowed for research purposes only.</p>

</body>

