<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Promoter analysis / ClusterBuster</h2>

<h3>Description</h3>

Retrieves promoter sequences and finds clusters of putative transcription factor binding sites in them. 
Transcription factor binding sites are searched for using the matrices from the
<a href="http://jaspar.cgb.ki.se/cgi-bin/jaspar_db.pl">JASPAR</a> library.

<h3>Parameters</h3>

<ul>
<li>Species (human, mouse, rat, drosophila, yeast) [human]</li>
<li>Promoter size (short, medium, long) [short] </li>
<li>Cluster score threshold [5]</li>
<li>Motif score threshold [6]</li>
<li>Expected distance between motifs in a cluster [35]</li> 
<li>Range for counting nucleotide frequencies [100]</li>
<li>Pseudocounts [0.375]</li>


</ul>

<h3>Details</h3>

<p>This tool retrieves upstream sequences for the specified genes and submits them to the ClusterBuster program. It needs to access the
chip-specific annotations, so if you have not specified the chiptype
during normalization of, e.g., Illumina data, it will not work. RefSeq IDs are used for retrieving 
promoter sequences constructed and annotated by UCSC genome browser staff. The same promoter sequences can be downloaded 
as a single FastA-formatted file from <a href="http://hgdownload.cse.ucsc.edu/downloads.html">UCSC Golden Path folder</a>. 
User can define how long promoter sequences are used for the analysis:</p>

<pre>
		Human		Mouse		Rat		Drosophila	Yeast
Short		1000 bp		1000 bp		1000 bp		1000 bp 	500 bp
Medium		2000 bp		2000 bp		2000 bp		2000 bp		1000 bp
Long		5000 bp		5000 bp		5000 bp		5000 bp		2500 bp
</pre>


<h3>Output</h3>

A text file detailing the cluster of TFBSs found, and their significance. The cluster are marked in the sequences 
with capital letters.

<h3>Reference</h3>
Frith et al (2003) Nuc Acids Res, 31(13):3666-8

</body>

