<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Bowtie for single-end reads</h2>

<h3>Description</h3>

<p>This tools aligns single-end reads to publicly available genomes or transcriptomes. If you would like other reference genomes to be added, please contact us.
If you would like to align reads against your own datasets, please use the tool "Bowtie against own genomes"

<h3>Parameters</h3>

<ul>
<li>Genome or transcriptome (human genome hg19, mouse genome mm9, rat genome rn4, mouse miRBase17) [mouse genome mm9]</li>
<li>Number of mismatches allowed (0, 1, 2, 3) [2]</li>
<li>Consider mismatches only in the seed region (yes, no) [no]</li>
<li>Length of the seed region (5-100) [28]</li>
<li>Allowed total of mismatch qualities (1-100) [70]</li>
<li>Quality value format used (Sanger, Illumina GA v1.3 or later) [Sanger]</li>
<li>How many best category hits is a read allowed to have (1, 2, no limit) [no limit]</li>
<li>Put multireads to a separate file (yes, no) [no]</li>
<li>How many valid alignments are reported per read (1, 2, 3) [1]</li>
<li>Put unaligned reads to a separate file (yes, no) [no]</li>
</ul>

<h3>Details</h3>


<p>Bowtie aligns reads to a reference sequence such as genome or transcriptome. There are two modes: mismatches are considered either throughout the read, 
or only in the user-defined left part of the read. In the latter case also quality values are taken into account. The Chipster implementation 
Results are sorted and indexed bam files, which are ready for viewing in the Chipster genome browser. 
Note that this Bowtie tool uses public genomes provided by Chipster. If you would like to align reads against your own datasets, please use the tool "Bowtie against own genomes".</p>


<p>After retrieving the promoter sequences, they are submitted to the Weeder program that finds common motifs
(putative transcription factor binding sites). Weeder can find motifs of two different sizes. The sizes are the same for all species. The length of the common motif and 
the number of allowed mismatches in the motifs varies with the setting:</p>

<pre>
			Length		Mismatches
Small			6		1
			8		2
Medium			10		3
<!--Large			12		4
Extra			6		1
			8		3
			10		4
			12		4-->
</pre>

<p>In addition, the user can define:</p>

<ul>
<li>whether both strands of DNA should be searched for the common motifs</li>
<li>how many times (none or once, several) the motif is expected to appear on every sequence</li>
<li>how many motifs to return</li>
<li>in how many of the sequences the motif should appear to be reported.</li>
</ul>

<h3>Output</h3>

<p>Tool returns a specified number of common motifs written to an HTML file. These can then be further analyzed to 
verify whether they are known transcription factor binding sites. </p>

<h3>Reference</h3>
Pavesi et al  (2004) Nuc Acids Res. Jul (W199-203) 


</body>

