<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Weeder</h2>

<h3>Description</h3>

<p>This tools retrieves promoter sequences for a set of genes and finds shared sequence motifs in them. Currently the tool works only for human, mouse, rat, drosophila and yeast genes.</p>

<h3>Parameters</h3>

<ul>
<li>Species (human, mouse, rat, drosophila, yeast) [human]</li>
<li>Promoter size (small, medium, large) [small]</li>
<li>Analyze strands (single, both) [single] </li>
<li>Motif appears more than once per sequence (yes, no) [no]</li>
<li>Number of motifs to return (1-100) [10]</li>
<li>Percentage of sequences the motif should appear in (1-100) [50]</li>
<li>Transcription factor binding site size (small, medium) [small]</li>
</ul>

<h3>Details</h3>


<p>This tool retrieves upstream sequences for the specified genes and submits them to the Weeder program. It needs to access the
chip-specific annotations, so if you have not specified the chiptype
during normalization of, e.g., Illumina data, it will not work. RefSeq IDs are used for retrieving 
promoter sequences constructed and annotated by UCSC genome browser staff. The same promoter sequences can be downloaded 
as a single FastA-formatted file from <a href="http://hgdownload.cse.ucsc.edu/downloads.html">UCSC Golden Path folder</a>. 
User can define how long promoter sequences are used for the analysis:</p>

<pre>
		Human		Mouse		Rat		Drosophila	Yeast
Small		1000 bp		1000 bp		1000 bp		1000 bp 	500 bp
Medium		2000 bp		2000 bp		2000 bp		2000 bp		1000 bp
Large		5000 bp		5000 bp		5000 bp		5000 bp		2500 bp
</pre>

<p>After retrieving the promoter sequences, they are submitted to the Weeder program that finds common motifs
(putative transcription factor binding sites). Weeder can find motifs of two different sizes. The sizes are the same for all species. The length of the common motif and 
the number of allowed mismatches in the motifs varies with the setting:</p>

<pre>
			Length		Mismatches
Small			6		1
			8		2
Medium			10		3
<!--Large			12		4
Extra			6		1
			8		3
			10		4
			12		4-->
</pre>

<p>In addition, the user can define:</p>

<ul>
<li>whether both strands of DNA should be searched for the common motifs</li>
<li>how many times (none or once, several) the motif is expected to appear on every sequence</li>
<li>how many motifs to return</li>
<li>in how many of the sequences the motif should appear to be reported.</li>
</ul>

<h3>Output</h3>

<p>Tool returns a specified number of common motifs written to an HTML file. These can then be further analyzed to 
verify whether they are known transcription factor binding sites. </p>

<h3>Reference</h3>
Pavesi et al  (2004) Nuc Acids Res. Jul (W199-203) 


</body>

