<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Utilities / Call SNPs and short INDELs for multiple individuals</h2>

<h3>Description</h3>

<p>Calls SNPs and short INDELs for multiple individuals. Individuals are identified from the SM tags in the @RG header lines. 
Individuals can be pooled in one alignment file. One individual can also be separated into multiple files.</p>

<h3>Parameters</h3>

<ul>
<li>Reference sequence (human hg19, mouse mm9, rat rn4) [human hg19]</li>
<li>Chromosome names in my BAM file look like (1, chr1) [chr1]</li>
<li>Platform (CAPILLARY, HELICOS, IONTORRENT, ILLUMINA, IONTORRENT, LS454, PACBIO, SOLID) [ILLUMINA]</li>
<li>Downgrading coefficient (Integer) [0]</li>
<li>Maximum read depth (Integer) [100]</li>
</ul>

<h3>Details</h3>

<p> 
You can provide your own reference sequence in FASTA format or choose one of the provided reference genomes. 
If you use the refrence genomes provided by Chipster, please indicate the style of chromosome naming in your BAM file (e.g. chr1 or 1).
</p>
<p>
Platform option specifies that indel candidates should be collected only from read groups with the @RG-PL tag set 
to that platform. Collecting indel candidates from reads sequenced by an indel-prone technology may affect the 
performance of indel calling.
</p>
<p>
You can provide a coefficient for downgrading mapping quality for reads containing excessive mismatches. Given a 
read with a phred-scaled probability q of being generated from the mapped position, the new mapping quality is 
about sqrt((INT-q)/INT)*INT. A zero value disables this functionality. Applying this option usually 
helps BWA-short but may not other mappers. The recommended value for BWA is 50.
</p>
<p> 
Maximum read depth should be adjusted to about twice the average read depth.
</p>


<h3>Output</h3>

<p>
Output is a VCF file. You can view VCF files in Chipster as text (which includes the header) or spreadsheet (which can be shorted). 
The location column in the spreadsheet can be used in Chipster genome browser to navigate quickly through a list of variants.
</p>

<h3>References</h3>

<p>This tool is based on the <a href = "http://samtools.sourceforge.net/">SAMtools package</a>. 
Please cite the article <a href = "http://www.ncbi.nlm.nih.gov/pubmed/19505943">The Sequence alignment/map (SAM) format and SAMtools</a> by Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N., Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing Subgroup (2009) Bioinformatics, 25, 2078-9. [PMID: 19505943]. 
</p>

</body>

