<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Exome-seq / Call SNPs and short INDELs for multiple individuals</h2>

<h3>Description</h3>

<p>Call SNPs and short INDELs for multiple individuals. </p>

<h3>Parameters</h3>

<ul>
<li>Platform (CAPILLARY, HELICOS, IONTORRENT, ILLUMINA, IONTORRENT, LS454, PACBIO, SOLID) [ILLUMINA]</li>
<li>Maximum read depth (Integer) [100]</li>
</ul>

<h3>Details</h3>

<p> 
You can provide your own reference sequence in FASTA format or choose one of the provided reference genomes. 

Individuals are identified from the SM tags in the @RG header lines. Individuals can be pooled in one alignment 
file. One individual can also be separated into multiple files. 

Platform option specifies that indel candidates should be collected only from read groups with the @RG-PL tag set 
to that platform. Collecting indel candidates from reads sequenced by an indel-prone technology may affect the 
performance of indel calling. 

Maximum read depth should be adjusted to about twice the average read depth.
</p>



<h3>Output</h3>

<p>
Output is a vcf file.
</p>


<h3>References</h3>

<p>This tool is based on the <a href = "http://samtools.sourceforge.net/">SAMtools package</a>. 
Please cite the article <a href = "http://www.ncbi.nlm.nih.gov/pubmed/19505943">The Sequence alignment/map (SAM) format and SAMtools</a> by Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N., Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing Subgroup (2009) Bioinformatics, 25, 2078-9. [PMID: 19505943]. 
</p>

</body>

