<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Utilities / Remove duplicate reads from BAM</h2>

<h3>Description</h3>

<p>
Given a BAM file, this tool removes reads which have identical external coordinates.
</p>

<h3>Parameters</h3>

<ul>
<li>Is the data paired end or single end reads (paired, single) [paired]</li>
</ul>

<h3>Details</h3>

<p>
You have to indicate if you have single end or paired end data. 
In the paired end mode the insert size (ISIZE) has to be correctly set in the BAM file. 
If multiple read pairs have identical coordinates, only the pair with highest mapping quality is kept. 
Note that this tool doesn't work for orphan reads or pairs which map to different chromosomes.
</p>


<h3>Output</h3>

<p>
Output is a BAM file.
</p>

<h3>References</h3>

<p>This tool is based on the <a href = "http://samtools.sourceforge.net/">SAMtools package</a>. 
Please cite the article <a href = "http://www.ncbi.nlm.nih.gov/pubmed/19505943">The Sequence alignment/map (SAM) format and SAMtools</a> by Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N., Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing Subgroup (2009) Bioinformatics, 25, 2078-9. [PMID: 19505943]. 
</p>

</body>

