<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title></title></head>
<body>

<h2>Statistics / Several group tests</h2>

<h3>Description</h3>

Compares the mean expression of every gene in the specified groups.

<h3>Parameters</h3>

<ul>
<li>Test type (empiricalBayes, ANOVA, Kruskal-Wallis) [ANOVA]</li>
<li>Method of multiple testing correction (none, Bonferroni, Holm, Hochberg, BH, BY) [BH]</li>
<li>P-value cut-off (0...1) [0.05]</li>
</ul>

<h3>Details</h3>

<p>Runs the specified test, and returns the genes that have a corrected p-value smaller than the specified 
p-value cut-off. If you want to get a p-value for every single gene in your data set, set the p-value cut-off to 1. This
should return a new gene list of equal length to the original one.</p>

<p>Multiple testing correction options are Bonferroni, Holm, and Hochberg for family-wise error rate (FWER) and
Benjamini-Hochberg and Benjamini-Yakutieri for false discovery rate (FDR). Of these Bonferroni is the most
conservative, returning the smallest number of significant results, and FDR-based adjustments are less
conservative, and return more significant results.</p>


<h3>Output</h3>

A text file containing the gene expression values and the p-value for the test.

<h3>References</h3>

<p>This tool uses Bioconductor package limma. Please cite the following articles, if you used empirical Bayes:</p>

<p>Smyth, G. K. (2004). Linear models and empirical Bayes methods for assessing
differential expression in microarray experiments. Statistical Applications in Ge-
netics and Molecular Biology, Vol. 3, No. 1, Article 3.</p>



</body>

