<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head><title>Chipster manual</title></head>
<body>

<h2>Analysis tools for next generation sequencing data</h2>

<p>Every tool is described in detail on its own help page. 
The help page is divided into four parts: General description, Parameters, Details on how the tool works, and Output 
where the results are described. The Parameters part contains the list of parameters, their possible values (1-100)
and their default values [10].</p>

<ol>
<li> <a href="#filtering">Filtering</a> </li>
<li> <a href="#quality">Quality control</a> </li>
<li> <a href="#utilities">Utilities</a> </li>
<li> <a href="#regions">Matching genomic regions</a> </li>
<li> <a href="#alignment">Alignment</a> </li>
<li> <a href="#RNA">RNA-seq</a> </li>
<li> <a href="#miRNA">miRNA-seq</a> </li>
<li> <a href="#ChIP">ChIP-seq</a> </li>
<li> <a href="#methyl">Methyl-seq</a> </li>
</ol>

<ul>
<b><a name="filtering">Filtering</a></b>
<li> <a href="ngs-filter-annotations.html">Filter by column term </a> </li>
<li> <a href="ngs-filter-results-column.html">Filter by column value </a> </li>
<br>
<br>
<b><a name="utilities">Utilities</a></b>
<li> <a href="bedtools-bamtobed.html">BEDTools bamToBed </a> </li>
<li> <a href="bedtools-fastafrombed.html">BEDTools fastaFromBed </a> </li>
<li> <a href="bedtools-mergebed.html">BEDTools mergeBed </a> </li>
<li> <a href="bedtools-slopbed.html">BEDTools slopBed </a> </li>
<li> <a href="bedtools-subtractbed.html">BEDTools subtractBed </a> </li>
<br>
<br>
<b><a name="regions">Matching genomic regions</a></b>
<li> <a href="bedtools-intersectbed.html">BEDTools intersectBed </a> </li>
<li> <a href="bedtools-closestbed.html">BEDTools closestBed </a> </li>
<li> <a href="bedtools-windowbed.html">BEDTools windowBed </a> </li>
<li> <a href="bedtools-coveragebed.html">BEDTools coverageBed </a> </li>
<li> <a href="bedtools-genomecoveragebed.html">BEDTools genomeCoverageBed </a> </li>
<li> <a href="bedtools-overlap.html">BEDTools overlap </a> </li>
<li> <a href="bedtools-pairtobed.html">BEDTools pairToBed </a> </li>
<li> <a href="bedtools-pairtopair.html">BEDTools pairToPair </a> </li>
<br>
<br>
<b><a name="alignment">Alignment</a></b>
<li> <a href="bowtie.html">Bowtie for single-end reads </a> </li>
<li> <a href="bowtie-paired-end.html">Bowtie for paired-end reads</a> </li>
<li> <a href="bowtie-with-index-building.html">Bowtie for single-end reads and own genome </a> </li>
<li> <a href="bowtie-paired-end-with-index-building.html">Bowtie for paired-end reads and own genome </a> </li>
<li> <a href="bwa.html">BWA for single-end reads </a> </li>
<li> <a href="bwasw.html">BWA-SW for single-end reads </a> </li>
<li> <a href="bwa-paired-end.html">BWA for paired-end reads </a> </li>
<li> <a href="bwa-with-index-building.html">BWA for single-end reads and own genome</a> </li>
<li> <a href="bwasw-with-index-building.html">BWA-SW for single-end reads and own genome</a> </li>
<li> <a href="bwa-paired-end-with-index-building.html">BWA for paired-end reads and own genome </a> </li>
<br>
<br>
<b><a name="ChIP">ChIP-seq</a></b>
<li> <a href="ngs-find-peaks-macs-one.html">Find peaks using MACS, treatment only </a> </li>
<li> <a href="ngs-find-peaks-macs-two.html">Find peaks using MACS, treatment vs. control </a> </li>
<li> <a href="ngs-find-motifs-jaspar.html">Find common motifs and match to Jaspar </a> </li>
<li> <a href="ngs-find-nearest-genes.html">Find the nearest genes for regions </a> </li>
<li> <a href="pathways-ngs-hyperg-go.html">GO enrichment for list of genes </a> </li>
<br>
<br>
<b><a name="methyl">Methyl-seq</a></b>
<li> <a href="convertBAMtoMEDIPS.html">Convert BAM file to MEDIPS input format </a> </li>
<li> <a href="combineMEDIPS.html">Combine MEDIPS input files </a> </li>
<li> <a href="MEDIPS_one-sample.html">MEDIPS -methylation analysis, treatment only </a> </li>
<li> <a href="MEDIPS_two-samples.html">MEDIPS -methylation analysis, treatment and control </a> </li>




      


<!--        <li> <a href=" "> </a> </li>        -->
</ul>

</body>

</html>