<html>
<head>
   <title>Chipster - Illumina tutorial</title>
</head>
<body>


<h2>Illumina tutorial I</h2>

<p>In this tutorial we will go through the following steps using Illumina gene expression data:
<ul>
<li>Data import</li>
<li>Preprocessing</li>
<li>Filtering</li>
<li>Statistical testing</li>
<li>Annotation</li>
<li>Exporting files</li>
</ul>


<h3>Biological question</h3>

<p>The dataset consists of two kinds of sperm cell samples. Five of the samples are normal sperm,
and eight are from teratozoospermic individuals. Teratozoospermia is a condition where most of the
sperm cells are severely malformed. The aim of the study is to identify the genes that are 
differentially expressed between these two different sperm types. The data set used in this tutorial is available from the GEO database with accession number 
<a href="http://www.ncbi.nlm.nih.gov/projects/geo/gds/gds_browse.cgi?gds=2696">GDS2696</a>.</p>


<h3>Data requirements</h3>

<p>The Illumina data is stored in separate files in GEO. One file represents one sample. Put all the datafiles in the same folder on, say, Desktop. In practise, it is easiest to read in a whole
folder at the same time.</p>


<h3>Data import</h3>

<p>To import the files to Chipster, go to menu File, and select Import Files:</p>

<img src="t3-02.jpg">

<p>Browse to the folder where the data is located:</p>

<img src="t3-03.jpg">

<p>There is a mix of .txt and .gz files in the folder. To display only the txt-files, select Microarray files
from the drop-down menu in the Open dialog box.</p>

<img src="t3-04.jpg">

<p>Select all the files by clicking on the first file, and the last file while keeping the Shift-key pressed down.
After selecting the files, click on the button Open to start the data import:</p>

<img src="t3-05.jpg">

<p>A new window opens, which allows you to define the actions performed for each file. You do not need to change anything, so click on OK.</p>

<img src="t3-06.jpg">

<p>Import Tool Opens. </p>

<img src="t3-07.jpg">

<p>Click on the 'Mark title row' button on the top.</p>

<img src="t3-08.jpg">

<p>Click on the title row on the table below to mark it. This just tells the program that the first
line of the file contains column names. These column names are then used to identify the column in later
stages of data import.</p>

<img src="t3-09.jpg">

<p>After marking the title row, click on the Next button on the bottom. This will get you to the next 
stage of data import.</p>

<img src="t3-10.jpg">

<p>Notice the buttons on the top row. For the current data, you would need to label at least one column
as Identifier and another column as Sample. In addition, one column could be labeled as flag, and
it can be used for automatically generating flags during the normalization phase.</p>

<p>To label a column, click on the button on the top row, and then the column in the table below.
First label column TargetID as Identifier. Next, label column AVG_Signal as sample. And last, label
column Detection as Flag. All other columns should remain unlabeled:</p>

<img src="t3-11.jpg">

<p>After this you've completed the data description phase, and can import the files to Chipster.
Click on the Finish button to accomplish this.</p>


<p>Once the data has been loaded, you'll see a list of 13 files appearing under the Dataset view. The same
set of files is also represented by a grey box with a text '13 files' appering in the Workflow View:</p>

<img src="t3-12.jpg">

<p>The next step would be to preprocess and normalize these data.</p>


<h3>Data preprocessing</h3>

<p>Click on the grey box representing the files in the Workflow view. This selects all the files:</p>

<img src="t3-13.jpg">

<p>Next, go to the Analysis tool list, and select the category Normalization. You'll see a list of 
preprocessing and normalization tools appearing:</p>

<img src="t3-14.jpg">

<p>Select the Illumina tool:</p>

<img src="t3-15.jpg">

<p>To view the current settings and to modify them, click on the button Edit parameters in the top-right
of the Chipster window. A list of parameters appears:</p>

<img src="t3-16.jpg">

<p>The current settings need to be modified to suit this data. As the Detection columns were exported
from the Import Tool, we can calculate flags for all genes. Change the setting Produce Flags to
yes. In addition you need to specify the chiptype. For this dataset it should be changed to Human-6v1.</p>

<img src="t3-17.jpg">

<p>After that press the Run-button to start the tool. A moving blue bar appears on the bottom of the Chipster window 
to indicate that a job is running:</p>

<img src="t3-18.jpg">

<p>It takes a couple of minutes to run the preprocessing and normalization, but once the results are ready 
they will appear under the Datasets and Workflow view. The normalized data set appears as a yellowish 
box in the Workflow view:</p>

<img src="t3-19.jpg">

<p>Before any real analysis, the phenodata of the dataset should be filled in. Phenodata is a description
of the experiment - what are the important questions that should be answered. Note the yellowish blob appearing
next to the preprocessed and normalized data set in the Workflow view - it is the phenodata. Click on the blob 
to select it:</p>

<img src="t3-20.jpg">

<p>Go to Data Visualization, and from the drop-down menu select the Phenodata editor:</p>

<img src="t3-21.jpg">

<p>This opens the Phenodata editor, where you should fill in the group column. Because the comparison of
normal sperm to malformed sperm was of principal interest in this experiment, we'll use the knowledge of
the sperm type to fill in the group column. All the samples from normal sperm are coded with 1, and
the samples from abnormal sperm with 2:</p>

<img src="t3-22.jpg">

<p>Once the phenodata column group is filled in, the exclamation mark disappears from the phenodata blob in 
the Workflow view:</p>

<img src="t3-23.jpg">

<p>Now you're ready to run some analysis.</p>


<h3>Quality control (optional)</h3>

<p>Illumina quality control is typically run on normalized data. To make the basic quality checks, select the 
normalized data in the Workflow view, and use Illumina -tool from the Quality control category:</p>

<img src="t3-24.jpg">

<p>It takes a few minutes to complete the quality control analysis. Once the analysis is ready, 
two new yellowish boxes appear under the Workflow view:</p>

<img src="t3-25.jpg">

<p>Select the left Prep-labeled box, and select Show Image from the visualization drop-down menu.
A density plot for all chips appears. After quantile normalization all chips have almost exactly the same
distribution of expression values, so the image is not very interesting.</p>

<img src="t3-26.jpg">

<p>The same applies to the box plot generated from the normalized data. That's the right-most Prep-box.
To view the image in its larger form, click on the Maximize button on the top bar of the Data visualization
area:</p>

<img src="t3-28.jpg">

<p>To get back to the original view, click on the Restore button on the top bar:</p>

<img src="t3-27.jpg">

<p>There are several more ways to quality control the normalization of the data. One of those
is non-metric multidimensional scaling. It's located under Statistics category with the name NMDS.
You don't need to modify any parameters, so just run the tool:</p> 

<img src="t3-29.jpg">

<p>Then visualize the result (the orange Stat-box). The relationships between the chips are visualized
in two dimensions in the NMDS plot. Notice the sample 13 that seems to deviate from the others 
especially on the Dimension 2 axis.</p>

<img src="t3-30.jpg">

<p>We can check the quality also by using hierarchical clustering. There is tool Dendrogram under the
Visualization category that does exactly that. You should modify the parameters to cluster chips, not
genes. After the run is ready, visualize the color-image. It is easy to see that the two groups
are not easily differentiated on the bases of all genes. The colored bar on the bottom tells each samples
group. The ones colored with blue are mostly group 2 samples. The sample 13 is the most distant from
all others, but not maybe deviation enough to warrant its removal from the dataset.</p>

<img src="t3-31.jpg">

<p>So we keep all the samples in the analysis, and move on to filtering.</p>



<h3>Filtering</h3>

<p>To filter the non-changing genes, select the normalized data set under the Workflow view:</p>

<img src="t3-32.jpg">

<p>One way to filter out the non-changing genes is to use filtering by standard deviation. The genes that show
the lowest standard deviation are those that are not displaying much changes between the normal and cancerous 
tissues. To filter the genes based on standard deviation, go to the tool category Preprocessing, and select the tool
Filter by standard deviation:</p>

<img src="t3-33.jpg">

<p>By default the tool filters out 99.7% of the genes:</p>

<img src="t3-34.jpg">

<p>After the filtering, a new data set is returned (the new Prep-box):</p>

<img src="t3-35.jpg">

<p>To see the genes that were retained in the filtered data, select the data set, and select Spreadsheet from the
drop down menu under the Data visualization.</p>

<p>The data is displayed as a spreadsheet, and the top row of the visualization says ' Showing 142 of 142 rows', so 
142 genes were retained after the filtering:</p>

<img src="t3-36.jpg">

<p>This filtered data can then be used for statistical testing.</p>


<h3>Statistical testing</h3>

<p>There are two groups in the data: normal sperm and abnormal sperm. The correct statistical test is the one
intended for two groups. To run this kind of test, go to the tool category Statistics, and select the tool Two
groups tests:</p>

<img src="t3-37.jpg">

<p>By default, the tool uses an empirical Bayes t-test for comparing the groups. This is a more sensitive test
than the standard t-test, and hence preferred. In addition, Benjamini and Hochberg's false discovery rate
is used for correcting for multiple tests, and a false discovery rate of 0.05 is used. It is advisable, at least
initially, to run the test using these settings:</p>

<img src="t3-38.jpg">

<p>After the analysis has finished an orange box appears in the Workflow view. If it is visualized
as a spreadsheet, it becomes apparent that 0 genes passed the statistical test:</p>

<img src="t3-39.jpg">

<p>After testing a few more settings, it fast become evident that most of the genes are not statistically 
significant at the usual level (0.05). So, we can calculate a p-value for every gene in the data set by
setting the p-value to 0, and turning the multiple testing correction off. This should return 142 genes
in the new dataset:</p>

<img src="t3-40.jpg">

<p>The second to last column noew lists the un-corrected p-values for all genes. To retain only the genes that
have a p-value of 0.2 or higher, we need to run the tool Extract genes using a p-value. It is located under the 
Utilities category:</p>

<img src="t3-41.jpg">

<p>Modify the settings so that the p-value threshold is 0.20, and the column that contains the p-values is the
p.adjusted column in the selected data set. Then run the tool:</p>

<img src="t3-42.jpg">

<p>P-value filtering returns 32 genes:</p>

<img src="t3-43.jpg">



<h3>Annotation</h3>

<p>Next, the 32 selected genes are annotated. To annotate the genes using the default Illumina annotations, 
select the tool category Annotation and the tool Affymetrix or Illumina gene list:</p>

<img src="t3-44.jpg">

<p>You can change the settings of the annotation script to include one column to display the p-values of the
genes and another one to display the fold changes of the genes between the two tissue types.</p>

<img src="t3-45.jpg">

<p>The tool returns an HTML page that contains the annotations for the genes. This page is displayed as a new
data set labeled Anno in the Workflow view. It can be visualized by double-clicking it:</p>

<img src="t3-46.jpg">
 
<p>You can click on the links on the page to open them in a www-browser. </p>


<h3>Exporting files</h3>

<p>In case you want to use the normalized or analyzed data sets in some other programs, you can easily export them.
Select the data set from the Workflow view, right-click on it, and select Export from the opening
menu:</p>

<img src="t3-47.jpg">

<p>Browse to the folder you want to export the data set, and press the Save button:</p>

<img src="t3-48.jpg">

<p>The same method can be used for exporting any other file from Chipster, including the images and the HTML page
with annotations.</p>


<h3>More advanced analyses</h3>
<p>The story continues in the <a href="tutorial4.html">Tutorial - part II</a>, where we will look into more details about 
annotations, KEGG pathways, GO ontology term enrichment, promoter analysis, and exporting the data to public databases.</p>


</body>
</html>
